File: packages/risk-analysis-data-pipeline/scraper/risk_pipeline/registry.py
"""Roster loading — the curated list of what gets scraped.

The roster lives in ``roster.json`` at the package root and is read by *both*
languages: TypeScript validates it with zod, Python with pydantic. Keeping one
file rather than a definition per language removes the class of bug where the
scraper collects a protocol the reader has never heard of, or vice versa.

Loading is strict: a malformed roster is a configuration error and should fail
loudly at startup rather than silently scrape nothing.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from .errors import ValidationError

__all__ = [
    "ChainTarget",
    "ProtocolTarget",
    "Roster",
    "load_roster",
]

#: Repo-relative location of the shared roster.
_ROSTER_PATH = Path(__file__).resolve().parents[2] / "roster.json"


@dataclass(frozen=True, slots=True)
class ChainTarget:
    """A chain to profile from L2Beat.

    Args:
        slug: URL-safe identifier, also the snapshot key stem.
        name: Display name.
        l2beat_path: L2Beat's own slug for the project page.

    """

    slug: str
    name: str
    l2beat_path: str


@dataclass(frozen=True, slots=True)
class ProtocolTarget:
    """A protocol whose governance forum to profile.

    Args:
        slug: URL-safe identifier, also the snapshot key stem.
        name: Display name.
        category: One of lending, dex, perps, yield, stablecoin, other.
        forum_url: The forum root, or ``None`` when the protocol has no forum.
        transport: What the capability probe found — one of ``discourse-json``,
            ``html-only``, ``blocked``, ``absent`` or ``unknown``.
        verified: Whether the transport has been confirmed by a live probe.
        note: Optional explanation, used when a protocol is deliberately skipped.

    """

    slug: str
    name: str
    category: str
    forum_url: str | None
    transport: str
    verified: bool
    note: str | None = None


@dataclass(frozen=True, slots=True)
class Roster:
    """The parsed roster.

    Args:
        version: Roster schema version.
        chains: Chains to profile.
        protocols: Protocols whose governance to profile.

    """

    version: str
    chains: list[ChainTarget]
    protocols: list[ProtocolTarget]

    def protocol(self, slug: str) -> ProtocolTarget | None:
        """Look up a protocol by slug.

        Args:
            slug: The protocol identifier.

        Returns:
            The target, or ``None`` when the slug is not in the roster.

        """
        return next((p for p in self.protocols if p.slug == slug), None)

    def scrapable_protocols(self) -> list[ProtocolTarget]:
        """Return protocols that can actually be fetched.

        Returns:
            Protocols whose transport is not ``absent``, i.e. those with a forum
            that the probe could reach (or has not yet ruled out).

        """
        return [p for p in self.protocols if p.transport != "absent"]


def load_roster(path: Path | None = None) -> Roster:
    """Load and validate the roster.

    Args:
        path: Override for the roster location. Defaults to the package root.

    Returns:
        The parsed roster.

    Raises:
        ValidationError: When the file is absent or malformed. A bad roster
            must stop the run rather than produce an empty sweep that looks
            successful.

    """
    resolved = path or _ROSTER_PATH

    try:
        raw = json.loads(resolved.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise ValidationError("roster", f"roster not found at {resolved}", None) from exc
    except json.JSONDecodeError as exc:
        raise ValidationError("roster", f"roster is not valid JSON: {exc}", None) from exc

    if not isinstance(raw, dict):
        raise ValidationError("roster", "roster root must be an object", raw)

    chains_raw = raw.get("chains")
    protocols_raw = raw.get("protocols")
    if not isinstance(chains_raw, list) or not isinstance(protocols_raw, list):
        raise ValidationError("roster", "roster must have `chains` and `protocols` arrays", None)

    chains = [
        ChainTarget(
            slug=str(entry["slug"]),
            name=str(entry["name"]),
            l2beat_path=str(entry["l2beatPath"]),
        )
        for entry in chains_raw
    ]

    protocols = [
        ProtocolTarget(
            slug=str(entry["slug"]),
            name=str(entry["name"]),
            category=str(entry["category"]),
            forum_url=str(entry["forumUrl"]) if entry.get("forumUrl") is not None else None,
            transport=str(entry.get("transport", "unknown")),
            verified=bool(entry.get("verified", False)),
            note=str(entry["note"]) if entry.get("note") is not None else None,
        )
        for entry in protocols_raw
    ]

    return Roster(
        version=str(raw.get("version", "0.0.0")),
        chains=chains,
        protocols=protocols,
    )

File: packages/risk-analysis-data-pipeline/src/registry.ts
/**
 * Roster loading — the curated list of what gets scraped.
 *
 * The roster lives in `roster.json` at the package root and is read by *both*
 * this module and the Python worker, which validates it with pydantic. Keeping
 * one file rather than a definition per language removes the class of bug where
 * the scraper collects a protocol the reader has never heard of, or the reverse.
 *
 * Loading is strict: a malformed roster is a configuration error and should fail
 * loudly rather than silently resolve to an empty sweep.
 */

import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { z } from 'zod';
import { ValidationError } from './errors.js';

/** Where the roster file lives, relative to this module's build output. */
const ROSTER_PATH = join(dirname(fileURLToPath(import.meta.url)), '..', '..', 'roster.json');

/** A chain to profile from L2Beat. */
export const ChainTargetSchema = z.object({
  slug: z.string().min(1),
  name: z.string().min(1),
  l2beatPath: z.string().min(1),
});
export type ChainTarget = z.infer<typeof ChainTargetSchema>;

/**
 * How a protocol's governance was confirmed reachable.
 *
 * `verified` records whether a live probe confirmed the transport, so a reader
 * can tell "not yet checked" apart from "checked and unavailable".
 */
export const PROTOCOL_TRANSPORTS = [
  'discourse-json',
  'html-only',
  'blocked',
  'absent',
  'unknown',
] as const;
export const ProtocolTransportSchema = z.enum(PROTOCOL_TRANSPORTS);
export type ProtocolTransport = z.infer<typeof ProtocolTransportSchema>;

/** A protocol whose governance forum to profile. */
export const ProtocolTargetSchema = z.object({
  slug: z.string().min(1),
  name: z.string().min(1),
  category: z.enum(['lending', 'dex', 'perps', 'yield', 'stablecoin', 'other']),
  /** Absent when the protocol has no reachable forum. */
  forumUrl: z.string().nullable(),
  transport: ProtocolTransportSchema,
  verified: z.boolean(),
  /** Why a protocol is excluded, when it is. */
  note: z.string().optional(),
});
export type ProtocolTarget = z.infer<typeof ProtocolTargetSchema>;

/** The parsed roster. */
export const RosterSchema = z.object({
  version: z.string().min(1),
  chains: z.array(ChainTargetSchema),
  protocols: z.array(ProtocolTargetSchema),
});
export type Roster = z.infer<typeof RosterSchema>;

/**
 * Load and validate the roster.
 *
 * @param path - Optional override for the roster location.
 * @returns The parsed roster.
 * @throws {ValidationError} When the file is absent or malformed. A bad roster
 *   must stop the caller rather than produce an empty result that looks like a
 *   successful sweep.
 */
export function loadRoster(path?: string): Roster {
  const resolved = path ?? ROSTER_PATH;

  let raw: unknown;
  try {
    raw = JSON.parse(readFileSync(resolved, 'utf8'));
  } catch (err) {
    const detail = err instanceof Error ? err.message : String(err);
    throw new ValidationError('roster', `could not read roster at ${resolved}: ${detail}`, null);
  }

  const parsed = RosterSchema.safeParse(raw);
  if (!parsed.success) {
    const issue = parsed.error.issues[0];
    throw new ValidationError(
      issue?.path.join('.') ?? 'roster',
      issue?.message ?? 'roster failed validation',
      raw,
    );
  }
  return parsed.data;
}

/**
 * Return the protocols whose governance can actually be fetched.
 *
 * Protocols the capability probe classified as `absent` or `blocked` are
 * excluded, because the roster already records why they are unavailable and
 * retrying them every sweep would spend time to relearn a known answer.
 *
 * @param roster - The roster to filter.
 * @returns The scrapable protocol targets.
 */
export function scrapableProtocols(roster: Roster): ProtocolTarget[] {
  return roster.protocols.filter((p) => p.transport === 'discourse-json');
}
